### PR TITLE
Compiler implementation

### PR DESCRIPTION
**Pull request by b-studios**

### CODE REVIEW

**Comment by b-studios** [comment]
@mm0821 I fixed the rendering bug and evidence now renders as expected:

image

**Comment by mm0821** [comment]
@b-studios I've drafted an explanation of lift inference. Maybe you can take a look if you find the time.

File: docs/implementation/lift-inference.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
+---
+layout: docs
+title: Lift Inference
+---
+
+# Lift Inference
+
+Programs in the Effekt surface language are translated to the underlying Core language.
```
Comment (b-studios) [inline docs/implementation/lift-inference.md:8]: Maybe we can mention that `Core` corresponds to `System Xi` in Brachthäuser 2020 and to `System C` in the 2022 OOPSLA paper?

File: docs/implementation/lift-inference.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+In order to efficiently compile programs in Core, the overhead incurred by using
+effect handlers should be minimized. In particular, the search for the correct handler
+at runtime should be avoided. This can be facilitated by making so-called lifting
+information explicit and using this information for a translation to iterated CPS.
```
Comment (b-studios) [inline docs/implementation/lift-inference.md:12]: That's correct, but maybe mention that Core already is in "capability-passing style"?

File: docs/implementation/lift-inference.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+operation `Exc1`. As is visible in the signature of `abort` (with an empty effect set) the
+effect is handled by the outer handler, even though `abort` is called inside of the inner
+handler. The result of running the program thus is `"aborted one"`.
+
+By clicking on "edit" the corresponding code in Core becomes visible. There the effect
+operations become explicit capabilities bound at the corresponding handlers. When `abort`
+is called, the `Exc1`-capability is executed and has to jump over the inner handler to its
+binding site, or to put it differently, it has to be lifted. At runtime this amounts to
+searching for the correct handler on the stack. By making explicit which capabilities
+have to be lifted to which handlers, such a runtime search can be rendered unnecessary.
```
Comment (b-studios) [inline docs/implementation/lift-inference.md:51]: Great! Ok, here it is mentioned that Core performs capability passing.

File: docs/implementation/lift-inference.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+
+### Regions and subregion evidence
+
+A structured way to do so is to introduce a region system and explicit subregion evidence.
```
Comment (b-studios) [inline docs/implementation/lift-inference.md:56]:
```suggestion
A structured way to make explicit which capability has to be lifted to which handler is to introduce a region system and explicit subregion evidence.
```

File: docs/implementation/lift-inference.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+      do Exc1()
+    } with Exc2 {
+      () => "aborted two"
+    }
+  } with Exc1 {
+    () => "aborted one"
+  }
+```
+
+By clicking on "edit" the corresponding code in the Lifted Core language is shown. There
```
Comment (b-studios) [inline docs/implementation/lift-inference.md:81]: Maybe mention our PLDI 2022 paper here?

File: docs/implementation/lift-inference.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+the call of `Exc1`, making it explicit that `Exc1` has to jump over the inner handler to
+run in the region of the outer handler.
+
+
+### Region and evidence inference
+
+Back to the original example. There the call to `Exc1` is within the definition of `abort`
+which is not inside the inner handler. Nevertheless, `abort` is called inside the inner
+handler and hence the call to `Exc1` only is only executed there, so it should again be
+provided with the evidence bound at the inner handler.
```
Comment (b-studios) [inline docs/implementation/lift-inference.md:93]: The above paragraph is a bit difficult to understand.

**Comment by b-studios** [comment]
Great thanks @mm0821. I left a few comments, but LGTM otherwise

**Comment by b-studios** [comment]
In general, it would also be great to have a compiler pipeline picture summarizing the different phases (linking lift inference) to your description. The other phases could easily be described later in separate documents.

Right now, the main description of the compiler pipeline can be found in https://github.com/effekt-lang/effekt/blob/master/effekt/shared/src/main/scala/effekt/Compiler.scala -- though I just noticed that the overview does not mention lift.

**Comment by mm0821** [comment]
I've drafted the picture and a description of the pipeline.
@b-studios Please take a look.

**Comment by b-studios** [comment]
Looks great, thanks!

**Pull request merged**